### PR TITLE
organize handlers into a map, try for nicer state handling in modals

### PR DESCRIPTION
**Pull request by mhuebert**
- Put handlers in a map - motivation being, that we can then generate handlers programmatically
- Proof of concept for modals driven by 'local state' stored in `private_metadata`. I think this will many more workflows tractable, eg. buttons that show/hide different elements within a modal, multi-step processes of any kind. The local-state can be modified by simple reducing functions (see `modal-updater`) or by interactive elements that report when their values change. Currently one still has to write all the reducing functions manually but I included a little example of how we could support a `:on-action` key to include these functions inline.

"On my machine" latency is acceptable and I'd guess our staging/prod servers will be even closer to Slack, will have to keep this in mind.

### CODE REVIEW

File: src/org/sparkboard/server/server.clj [hunk 1]
```diff
@@ -523,13 +390,213 @@
     (case action
       :link-account! (link-account! context params))))
 
+(defn action-value [action]
```
Comment (daveliepmann) [inline src/org/sparkboard/server/server.clj:393]: What's the expected data model of an incoming `action`?

Comment (daveliepmann) [reply src/org/sparkboard/server/server.clj:393]: I'm thinking that might make sense in a docstring or `:pre` condition

File: src/org/sparkboard/server/server.clj [hunk 2]
```diff
@@ -523,13 +390,213 @@
...
+                    [:actions
+                     [:button {:modal/action-id "counter+"
+                               :modal/on-action #(update % :n inc)}
+                      (str "Counter: " (:n state))]]]))
+  ;; this would return
+  {:block_actions/my-modal-open ...
+   :block_actions/my-modal-counter+ ...})
+;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+(def slack-api-handlers
```
Comment (daveliepmann) [inline src/org/sparkboard/server/server.clj:454]: Putting events in one place (but distinguishable by namespace) works quite well here

File: src/org/sparkboard/server/server.clj [hunk 3]
```diff
@@ -523,13 +390,213 @@
...
+                          (let [[action-id action-id-suffix] (-> payload :actions first :action_id
+                                                                 (str/split (re-pattern screens/action-id-separator)))]
+                            [(keyword payload-type action-id)
+                             context
+                             payload
+                             {:action-id-suffix action-id-suffix
+                              :modal! (fn
+                                        ([blocks]
+                                         (log/info :modal! (hiccup/->blocks blocks))
+                                         (slack/web-api "views.open"
```
Comment (daveliepmann) [inline src/org/sparkboard/server/server.clj:583]: It may make sense to factor out this fn

File: src/org/sparkboard/server/slack/core.clj [hunk 4]
```diff
@@ -40,7 +40,7 @@
         rsp (json->clj (.body (.send clnt request (HttpResponse$BodyHandlers/ofString))))]
     (log/debug "[web-api] GET rsp:" rsp)
     (when-not (:ok rsp)
-      (throw (ex-info (str "web-api failure: slack/" family-method) {:rsp rsp :config config})))
+      (throw (ex-info (str "web-api failure: slack/" family-method) {:rsp rsp :config config :query-map query-map})))
```
Comment (daveliepmann) [inline src/org/sparkboard/server/slack/core.clj:43]: 👍  thanks

File: src/org/sparkboard/server/slack/screens.clj [hunk 5]
```diff
@@ -162,6 +174,22 @@
        :options [{:value "collect-in-thread"
                   :text [:md "Collect responses in a thread"]}]}]]]))
 
+(defn state-test-modal [context payload state]
```
Comment (daveliepmann) [inline src/org/sparkboard/server/slack/screens.clj:177]: I guess this is temporary?

File: src/org/sparkboard/server/server.clj (hunk 1 above)
Comment (mhuebert) [reply src/org/sparkboard/server/server.clj:393]: https://api.slack.com/reference/interaction-payloads/block-actions has some info. I haven't  found any good reference and am just expanding this function as I use more types.

> I'm thinking that might make sense in a docstring or :pre condition
What is this referring to? GH isn't showing me the context of this comment

File: src/org/sparkboard/server/server.clj (hunk 2 above)
Comment (mhuebert) [reply src/org/sparkboard/server/server.clj:454]: I am thinking so too. I started with a nested structure but it's much easier to work with this as a flat map.

File: src/org/sparkboard/server/slack/screens.clj (hunk 5 above)
Comment (mhuebert) [reply src/org/sparkboard/server/slack/screens.clj:177]: yeah, this is just a little test modal to see how the local-state stuff behaves.

File: src/org/sparkboard/server/server.clj (hunk 3 above)
Comment (mhuebert) [reply src/org/sparkboard/server/server.clj]: I've removed this and added some view helpers in the slack ns, which rely on a new dynamic var: https://github.com/sparkboard/sparkboard/pull/45/commits/e629fa67a5954312cc4d250d8dd35df2e3507e5d#diff-4fb98734dd17bcb63fac2abb0966053cR97

also added a 2-arity to web-api for auto-filling the token.

**Comment by mhuebert** [comment]
Some more updates..
- moving payload/event into context, to think less about what arities various functions have. I'm finding it easier to remember `:slack/event` than "what is the 2nd argument passed to this function"
- added examples of all the interactive component types (& figured out how to read their action-values properly)
- moved the reply ref out of the action-id and into block-id
- experimented with removing update-requests after reply. this looked weird, so I left a reply in place, with a permalink to where the update was posted. another alternative would be to update (/re-render) the existing message and replace the button with a thank-you. discovered slack's own weird [link format](https://github.com/sparkboard/sparkboard/pull/45/commits/c4992cb37fd33bd5d1e650cab04501d3a1b05417#diff-8e1cef957fe8b7cf7818171dcd9e1210R117)
- moved the modal utilities into their own namespace

This has turned into quite a bit of refactoring so we should probably merge it. I'd like to touch base with the OG folks soon to get their feedback on current status.

File: src/org/sparkboard/server/server.clj (hunk 1 above)
Comment (daveliepmann) [reply src/org/sparkboard/server/server.clj]: I meant a docstring/`:pre` condition to document the shape/keys of an incoming `action` in this `action-value` function.

**Pull request merged**